### PR TITLE
Add cache array for type weights

### PR DESCRIPTION
**Pull request by kampersanda**
I added a cache array that stores weight results for all type sequence patterns. When type_window_size = 3, the array consumes only 1 MiB and can accelerate the tokenization by bypassing pattern matching.

For example, with my environment (GCP, n1-standard-4), the prediction time for `kyoto-train.ja` with `jp-0.4.7-6` model was improved from 4.26 sec to 3.91 sec.

However, this approach uses a large amount of memory for a large type_window_size. So, it is enabled only if type_window_size <= 3. The runtime switching for an input model is implemented with Enum dispatch.

### CODE REVIEW

File: vaporetto/src/lib.rs [hunk 1]
```diff
@@ -41,6 +41,8 @@ mod trainer;
 #[cfg(feature = "kytea")]
 mod kytea_model;
 
+mod type_predictor;
```
Comment (vbkaisetsu) [inline vaporetto/src/lib.rs:44]: Could you move this to the next of `mod sentence`?

File: vaporetto/src/predictor.rs [hunk 2]
```diff
@@ -136,30 +139,7 @@ impl Predictor {
     }
 
     fn add_type_ngram_scores(&self, sentence: &Sentence, start: usize, ys: &mut [ScoreValue]) {
```
Comment (vbkaisetsu) [inline vaporetto/src/predictor.rs:141]: This function is no longer necessary.
Could you call `type_predictor.add_scores()` directly?

File: vaporetto/src/type_predictor.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
+use crate::model::ScoreValue;
+use crate::sentence::Sentence;
+use daachorse::DoubleArrayAhoCorasick;
+
+pub enum TypePredictor {
```
Comment (vbkaisetsu) [inline vaporetto/src/type_predictor.rs:5]: I think `TypeScorer` is better.
This name seems that the function predicts types.

File: vaporetto/src/type_predictor.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+pub struct TypePredictorCache {
+    scores: Vec<ScoreValue>,
+    window_size: usize,
+    sequence_mask: usize,
+}
+
+impl TypePredictorCache {
+    pub fn new(
+        pma: DoubleArrayAhoCorasick,
```
Comment (vbkaisetsu) [inline vaporetto/src/type_predictor.rs:86]: Currently, `TypePredictorCache` is built from the PMA.
I think this is complicated.
Why don't you use `model.types` directly?

Comment (kampersanda) [reply vaporetto/src/type_predictor.rs:86]: To build the cache array, the current implementation generates all possible sequences and looks up n-grams in each sequence. For the lookup, it is necessary to use some index, and I used PMA straightforwardly.

File: vaporetto/src/type_scorer.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        let all_sequences = ALPHABET_SIZE.pow(sequence_size as u32);
+
+        let mut sequence = vec![0u8; sequence_size];
+        let mut scores = vec![0 as ScoreValue; all_sequences];
+
+        for (i, score) in scores.iter_mut().enumerate() {
+            if !Self::seqid_to_seq(i, &mut sequence) {
+                continue;
+            }
+            let mut y = 0;
```
Comment (vbkaisetsu) [inline vaporetto/src/type_scorer.rs:100]: Could you fix errors on the CI?
```
$ cargo test -p vaporetto --no-default-features
...
error[E0277]: cannot add-assign `f64` to `{integer}`
   --> vaporetto/src/type_predictor.rs:102:19
    |
102 |                 y += weights[m.pattern()][sequence_size - m.end()];
    |                   ^^ no implementation for `{integer} += f64`
    |
    = help: the trait `AddAssign<f64>` is not implemented for `{integer}`

error[E0308]: mismatched types
   --> vaporetto/src/type_predictor.rs:104:22
    |
104 |             *score = y;
    |                      ^ expected `f64`, found integer
```

```suggestion
            let mut y = ScoreValue::default();
```

File: vaporetto/src/type_predictor.rs (hunk 4 above)
Comment (vbkaisetsu) [reply vaporetto/src/type_predictor.rs]: Okay.

File: vaporetto/src/type_scorer.rs (hunk 5 above)
Comment (kampersanda) [reply vaporetto/src/type_scorer.rs]: Fixed (I'm sorry I forgot to use Commit suggestion)

**Review by vbkaisetsu — APPROVED** [review]

**Pull request merged**